### PR TITLE
8251942: PrintStream specification is not clear which flush method is automatically invoked

### PR DESCRIPTION
**Pull request by bplb**
Please review this minor change to the specification of `java.io.PrintStream`. The longstanding behavior for flushing is to invoke the `flush()` method of the underlying `OutputStream` rather than its override but this was not made explicit in the specification.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8251942](https://bugs.openjdk.java.net/browse/JDK-8251942): PrintStream specification is not clear which flush method is automatically invoked

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2926/head:pull/2926`
`$ git checkout pull/2926`

To update a local copy of the PR:
`$ git checkout pull/2926`
`$ git pull https://git.openjdk.java.net/jdk pull/2926/head`

### CODE REVIEW

File: src/java.base/share/classes/java/io/PrintStream.java [hunk 1]
```diff
@@ -39,19 +39,19 @@
...
  * internal flag that can be tested via the {@code checkError} method.
  * Optionally, a {@code PrintStream} can be created so as to flush
- * automatically; this means that the {@code flush} method is
- * automatically invoked after a byte array is written, one of the
- * {@code println} methods is invoked, or a newline character or byte
- * ({@code '\n'}) is written.
+ * automatically; this means that the {@code flush} method of the underlying
+ * output stream is automatically invoked after a byte array is written, one
+ * of the {@code append}, {@code print}, or {@code println} methods is invoked,
+ * or a newline character or byte ({@code '\n'}) is written.
```
Comment (dfuch) [inline src/java.base/share/classes/java/io/PrintStream.java:45]: Though not wrong, this was a bit surprising. If I'm not mistaken in the case of `print` and `append` the `flush` method will be called by `write` only if the CharSequence (or String) contains a `\n` character. However, the complex layering where different output stream wrap themselves like Russian dolls (I'm talking about the use of `textOut` and `charOut` here) means that calling `print` or `append` eventually ends up in a call to `write(byte[], int, int)` on the PrintStream which causes a flush() to occur on the underlying stream. So in the case that the String contains a `\n` then flush will be called twice :-) 

I wonder how much of this is exposing arcane implementation details...

**Comment by dfuch** [comment]
> Yes, I noticed that as well. I didn?t think it was worth complicating things for the purpose of this issue to address it.

I guess what I was trying to ask is whether we should actually specify that `print` and `append` call `flush` - as this seems to be a side effect of some optimization.
Maybe we should say that the implementation ensure that flush is called when writing a byte array or when  a newline character or byte ({@code '\n'}) is written - but might call it in additional unspecified circumstances?

**Comment by bplb** [comment]
I think you are correct. In the second commit I scaled back the change to the class level specification. I left out any mention of "unspecified circumstances."

**Comment by bplb** [comment]
Is there any further comment on this request? Thanks.

**Review by dfuch — APPROVED** [review]
Looks good to me Brian!

**Review by AlanBateman — APPROVED** [review]

**Comment by bplb** [comment]
CSR filed: https://bugs.openjdk.java.net/browse/JDK-8263835.

**Pull request closed without merging**